[TT-1326] Update Solidty Foundry pipeline with Slither

Tofel opened the pull request:
* top level checks config
* add Slither for modified files, excluding tests (in the future only for diff lines)
* run code coverage for eligible products
* run Forge fmt for eligible products

Running everything, when `shared` file changed: https://github.com/smartcontractkit/chainlink/actions/runs/10200971215
Running only one product: https://github.com/smartcontractkit/chainlink/actions/runs/10201139248

Tofel committed:
More univeral lcov prunning

Tofel committed:
update Shared code cov

Tofel committed:
exclude deleted files from Slither

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -11,28 +11,55 @@ jobs:
+        product:
+          # 98.5 is the aspirational code coverage once we migrate all tests to Foundry
+          # product that have that minimum coverage of 98.5 in the matrix below are currently excluded from the check
+          - name: automation
```
Is it possible to move this to the top of the file, and also add the info if products have a gas snapshot, and if they use forge fmt? That would be a good single source of truth for all CI in this file

Tofel replied:
this is a really cool idea!

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -11,28 +11,55 @@ jobs:
     runs-on: ubuntu-latest
     outputs:
-      changes: ${{ steps.changes.outputs.src }}
+      src_changes: ${{ steps.changes.outputs.src }}
```
I think it makes sense to check for changes in the folders per product. Right now we would be running coverage for all products if any single one changes a line. Ideally, CI only runs for affected products. 

Any changes to `vendor` or `shared` would trigger a full run, don't think we can get around that

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -47,127 +74,223 @@ jobs:
 
-      - name: Report code coverage
-        if: ${{ contains(fromJson('["ccip"]'), matrix.product) && needs.changes.outputs.changes == 'true' }}
+      - name: Report code coverage for ${{ matrix.product.name }}
```
Would this not spam a github comment for every product? Would it be possible to maybe combine them into a single one? Less of a problem when we only run tests/cov for changed products but would still be kinda annoying getting coverage spam in comments. 

Imo it doesn't have to be reported at all, people can just look at failing CI

Tofel replied:
disabled, since there was a race condition any way; matrix job that executed last was the one that set the comment (and that was CCIP)

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -47,127 +74,223 @@ jobs:
+    needs: [ changes ]
+    name: Check formatting for modified files
+    if: needs.changes.outputs.sol_changes == 'true'
+    runs-on: ubuntu-22.04
```
Only CCIP code uses forge fmt

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -47,127 +74,223 @@ jobs:
+          FOUNDRY_PROFILE: ${{ matrix.product.name }}
         run: |
-          forge fmt --check
+          SOL_FILES="${{ needs.changes.outputs.sol_files }}"
```
Do we need this over the extremely simple check we had before? This is lightning fast

Tofel committed:
use single source of truth for all Solidity Foundry jobs

Tofel committed:
fix json

Tofel committed:
compact output with jq

Tofel committed:
fix condition for fmt

Tofel committed:
try to scope tests to changes

Tofel committed:
move matrix check to step level

Tofel committed:
fix outputs path

Tofel committed:
trigger

Tofel committed:
test with Automation change

Tofel committed:
try with shared

Tofel committed:
run fmt also if any sol files were modified

Tofel committed:
fix job name in collect metrics

Tofel committed:
trigger pipeline only for localised change + update changes info

Tofel committed:
add changeset

Tofel committed:
remove test change

RensR commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -47,127 +110,236 @@ jobs:
-    # passing required check for PRs that don't have filtered changes.
     steps:
       - name: Checkout the repo
+        if: ${{ (contains(fromJson(needs.changes.outputs.all_changes), matrix.product.name)
```
Probably don't need to format based on changes in `shared`

Tofel committed:
do not run forge fmt if shared contracts have changed

Tofel committed:
Merge branch 'develop' into tt_1326_update_Sol_Foundry

Tofel committed:
try artifact pipeline by hijacking hardhat

RayXpub commented on contracts/.slither.config-pr.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "filter_paths": "(openzeppelin|mocks/|test/|tests/|testhelpers)",
+  "detectors_to_exclude": "pragma"
```
suggestion: I think it might be useful to have a config file per product so that each one can have custom configs. For CCIP for example, we might want to add additional detectors to exclude such as:
- solc-version
- naming-convention
- assembly
- reentrancy-events
- timestamp
- calls-loop
- unused-return
- ...

And the above might not be applicable to all products.

Tofel replied:
added

momentmaker commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -6,39 +6,102 @@ env:
+              - 'contracts/src/v0.8/l2ep/**/*.sol'
+            liquiditymanager:
+              - 'contracts/src/v0.8/liquiditymanager/**/*.sol'
+            ll-feeds:
```
should be `llo-feeds` no?

Tofel replied:
good catch!

momentmaker commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -6,39 +6,102 @@ env:
+        id: define-matrix
+        shell: bash
+        run: |
+          cat <<EOF > matrix.json
```
maybe create a setting.json file?
https://www.jitsejan.com/use-github-actions-with-json-file-as-matrix

Tofel replied:
I thought about it, but decided that locality of action is more important, so that you see everything in a single place

momentmaker commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -6,39 +6,102 @@ env:
       - uses: dorny/paths-filter@de90cc6fb38fc0963ad72b210f1f284cd68cea36 # v3.0.2
         id: changes
         with:
+          list-files: 'shell'
```
I think you can use `csv` so you don't have to modify in the latter steps for parsing files.

```
    # Enables listing of files matching the filter:
    #   'none'  - Disables listing of matching files (default).
    #   'csv'   - Coma separated list of filenames.
    #             If needed, it uses double quotes to wrap filename with unsafe characters.
    #   'json'  - File paths are formatted as JSON array.
    #   'shell' - Space delimited list usable as command-line argument list in Linux shell.
    #             If needed, it uses single or double quotes to wrap filename with unsafe characters.
    #   'escape'- Space delimited list usable as command-line argument list in Linux shell.
    #             Backslash escapes every potentially unsafe character.
```

Tofel committed:
Merge branch 'develop' into tt_1326_update_Sol_Foundry

Tofel committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

CR changes + test them

# This is the commit message #2:

use shell array

Tofel committed:
CR changes + test them

use shell array

init array

CR changes + test them

use shell array

init array

Tofel committed:
remove test files

Tofel committed:
do not run Slither for test files

Tofel committed:
do not run fmt if test files were modified

Tofel committed:
remove unused config file

Tofel committed:
Merge branch 'develop' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
restore old Hardhat pipeline

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
add missing transmission setup

RayXpub commented on contracts/scripts/ci/generate_slither_report.sh:
```diff
@@ -0,0 +1,80 @@
+
+    SLITHER_OUTPUT_FILE="$TARGET_DIR/$(basename "${FILE%.sol}")-slither-report.md"
+
+    output=$(slither --config-file "$CONFIG_FILE" "$FILE" --checklist --markdown-root "$REPO_URL" --fail-none $SLITHER_EXTRA_PARAMS)
```
suggestion: I am thinking we might want to fail on high impact findings here and if it is a false positive then we update the config file to exclude the detector? one downside I see is if a detector should be ignored only on some files. cc @RensR

Tofel replied:
up to you guys, my idea was to make it purely informative for now, but you know best what you need :-)

RensR replied:
Maybe nice to follow up, but this is fine for now

Tofel committed:
fix basic info condition, join 2 steps into 1, define higher-level coverage exclusions

Tofel committed:
define actions for installing Slither and solc-select

Tofel committed:
run all tests also if package.json changes; run them on all non_src changes

Tofel committed:
add action for validating whether all Slither reports and UML diagrams were generated

Tofel committed:
fetch origin in validation action

Tofel committed:
compare with HEAD in validate action

Tofel committed:
compare with origin in validation action

Tofel committed:
handle both csv and shell arrays in the validation action

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
update artifact pipeline with new actions

Tofel committed:
Merge branch 'develop' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
fix workflow after tests

Tofel committed:
fix how validation actions works with commits

Tofel committed:
Merge branch 'develop' into tt_1326_update_Sol_Foundry

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
treat shared as any other product

Tofel committed:
small fixes

chainchad commented on .github/actions/setup-slither/action.yaml:
```diff
@@ -0,0 +1,10 @@
+      shell: bash
+      run: |
+        python -m pip install --upgrade pip
+        pip install slither-analyzer
```
Might be good to pin version this in a requirements.txt file or directly inline here like ` pip install slither-analyzer==0.10.3`.

chainchad commented on .github/actions/setup-solc-select/action.yaml:
```diff
@@ -0,0 +1,32 @@
+    - name: Install solc-select and solc
+      shell: bash
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y python3-pip
```
```suggestion
```

pip3 should already be installed, see the list of installed packages on the [ubuntu-22.04 runner image](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md).

chainchad commented on contracts/scripts/ci/modify_remappings.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
```suggestion
#!/usr/bin/env bash
```

chainchad commented on contracts/scripts/ci/modify_remappings.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
```
```suggestion
set -euo pipefail
```

chainchad replied:
I would suggest adding this to all bash scripts.

chainchad commented on contracts/scripts/ci/generate_slither_report.sh:
Where are you calling this script from?

chainchad commented on .github/workflows/solidity-foundry.yml:
```diff
@@ -47,127 +119,247 @@ jobs:
+      - name: Install semver
+        shell: bash
+        run: |
+          npm install -g semver
```
I believe the previous `setup-nodejs` action will install the dependencies from `./package.json` which already includes this `semver` package. Can probably remove this step.

Tofel committed:
apply CR changes

chainchad commented on .github/actions/setup-solc-select/action.yaml:
```diff
@@ -15,7 +15,6 @@ runs:
       shell: bash
       run: |
         sudo apt-get update
```
Can probably get rid of this apt update.

Tofel committed:
remove special handling for deleted files

Tofel committed:
remove apt-get update

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
use only dorny/paths

Tofel committed:
remove unused input

Tofel committed:
CR changes: use dorny/paths with quantifier, move scope validation to an action, remove whitespaces

Tofel committed:
fix workflow

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
fail bash scripts on erors

Tofel committed:
add set -euo pipefail to bash scripts

Tofel committed:
define action to detect foundry version

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
fix select solc version script, better slither report output

Tofel committed:
checkout repo

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
add id

Tofel committed:
Merge branch 'tt_1326_update_Sol_Foundry' into tt_1325_solidity_pipelines_genhtml

Tofel committed:
Merge pull request #14015 from smartcontractkit/tt_1325_solidity_pipelines_genhtml
